src/Build/Construction/ProjectImportElement.cs

```diff
@@ -60,144 +95,341 @@ public string Project
+            _ => throw CreateUnknownSourceException()
+        };
+
+        internal SdkReferenceWithOrigin? SdkReferenceWithOrigin => _sdkReferenceSource switch
+        {
+            ISdkReferenceMutableSource mutableSource => ComputeSdkReferenceWithOrigin(mutableSource),
+            SdkReferenceConstantSource constantSource => constantSource.SdkReference,
+            _ => throw CreateUnknownSourceException()
+        };
+
+        private SdkReferenceOrigin? SdkReferenceOrigin => _sdkReferenceSource switch
+        {
+            ISdkReferenceMutableSource mutableSource => ComputeSdkReferenceOrigin(mutableSource),
+            SdkReferenceConstantSource constantSource => constantSource.SdkReference.Origin,
+            _ => throw CreateUnknownSourceException()
+        };
 
         /// <summary>
-        /// Creates an unparented ProjectImportElement, wrapping an unparented XmlElement.
-        /// Validates the project value.
-        /// Caller should then ensure the element is added to a parent
+        ///     Creates an unparented ProjectImportElement, wrapping an unparented XmlElement.
+        ///     Validates the project value.
+        ///     Caller should then ensure the element is added to a parent
         /// </summary>
         internal static ProjectImportElement CreateDisconnected(string project, ProjectRootElement containingProject)
         {
             XmlElementWithLocation element = containingProject.CreateElement(XMakeElements.import);
-            return new ProjectImportElement(element, containingProject) {Project = project};
+            return new ProjectImportElement(element, containingProject)
+            {
+                Project = project
+            };
         }
 
         /// <summary>
-        /// Creates an implicit ProjectImportElement as if it was in the project.
+        ///     Creates an implicit ProjectImportElement as if it was in the project.
         /// </summary>
-        /// <returns></returns>
         internal static ProjectImportElement CreateImplicit(
             string project,
             ProjectRootElement containingProject,
             ImplicitImportLocation implicitImportLocation,
-            SdkReference sdkReference,
+            ISdkReferenceSource? sdkReferenceSource,
             ProjectElement originalElement)
         {
             XmlElementWithLocation element = containingProject.CreateElement(XMakeElements.import);
-            return new ProjectImportElement(element, containingProject)
+            return new ProjectImportElement(element, null, containingProject, sdkReferenceSource)
             {
                 Project = project,
-                Sdk = sdkReference.ToString(),
                 ImplicitImportLocation = implicitImportLocation,
-                SdkReference = sdkReference,
                 OriginalElement = originalElement
             };
         }
 
+#nullable restore
```

Comment:
Since I don't know the proper nullability of the following 2 methods, I've decided to skip annotating these (the same goes for the ones in `ProjectSdkElement`).